src/ape_cache/query.py

```diff
@@ -0,0 +1,169 @@
+                number_of_rows = q.rowcount
+
+        except Exception as err:
+            # Note: If any error, skip the data from the cache and continue to
+            #       query from provider.
+            logger.debug(err)
+            number_of_rows = 0
+
+        # NOTE: Assume 200 msec to get data from database
+        time_to_get_cached_records = 200 if number_of_rows > 0 else 0
+        # NOTE: Very loose estimate of 0.75ms per block
+        time_to_get_uncached_records = int(
+            ((query.stop_block - query.start_block) / query.step - number_of_rows) * 0.75
+        )
+        return time_to_get_cached_records + time_to_get_uncached_records
+
+    @singledispatchmethod
+    def perform_query(self, query: QueryType) -> pd.DataFrame:  # type: ignore
+        raise QueryEngineError(f"Cannot handle '{type(query)}'.")
+
+    @perform_query.register
+    def perform_block_query(self, query: BlockQuery) -> pd.DataFrame:
+        try:
+            with self.engine.connect() as conn:
+                q = conn.execute(
+                    text(
+                        """
+                        SELECT :columns
+                        FROM blocks
+                        WHERE blocks.number >= :start_block
+                        AND blocks.number <= :stop_block
+                        AND blocks.number mod :step = 0
+                        """
+                    ),
+                    columns=",".join(query.columns),
+                    start_block=query.start_block,
+                    stop_block=query.stop_block,
+                    step=query.step,
+                )
+                cached_records = pd.DataFrame(columns=query.columns, data=q.fetchall())
+
+        except Exception as err:
+            # Note: If any error, skip the data from the cache and continue to
+            #       query from provider.
+            logger.debug(err)
+            cached_records = pd.DataFrame(columns=query.columns)
+
+        if len(cached_records) == math.ceil((query.stop_block - query.start_block) / query.step):
+            return cached_records
+
+        blocks_iter = map(
+            self.provider.get_block,
+            # NOTE: the range stop block is a non-inclusive stop.
+            #       Where as the query method is an inclusive stop.
+            range(query.start_block + len(cached_records), query.stop_block + 1, query.step),
+        )
+        block_dicts_iter = map(partial(get_columns_from_item, query), blocks_iter)
+        return pd.concat(
+            [
+                cached_records,
+                pd.DataFrame(columns=query.columns, data=block_dicts_iter),
+            ]
+        )
+
+    def update_cache(self, query: QueryType, result: pd.DataFrame):
+        # TODO: Add handling of having primary key and potentially
+        #  updating table with certain columns
+        if set(result.columns) != set(query.all_fields()):
+            return  # We do not have all the data to update the database
```

Comment:
This is here to prevent inserting partially-defined rows into the database

With more advanced handling of detecting queries for rows that contain null data, we can perhaps allow partial inserts, but it would have to query when asking for data it hasn't cached yet for given rows